### PR TITLE
Factorise the common parts in keystore implementations

### PR DESCRIPTION
**Pull request by CedricCabessa**
remove duplication
The real code is moved to function in store.go
memory.go and redis.go do the access layer thing and call those functions

### CODE REVIEW

File: pkg/keystore/redis.go [hunk 1]
```diff
@@ -85,327 +77,112 @@ func (s *RedisKeystore) Reset(id uuid.UUID) error {
...
-
-			addrInfo := AddressInfo{
-				Address:    addr,
-				Derivation: path,
-				Change:     change,
-			}
-
-			addrs = append(addrs, addrInfo)
+	for _, meta := range changedMetas {
+		if err := set(s.db, id.String(), meta); err != nil {
```
Comment (CedricCabessa) [inline pkg/keystore/redis.go:123]: totally not sure here
We are overwritting the same key again? It was present in the original code, but in a different form

**Comment by CedricCabessa** [comment]
noob go question: go.mod and go.sum have changed locally.
Why? is it mage or the go compiler? I didn't ask for anything :-)
What I am supposed to do with that?

**Comment by CedricCabessa** [comment]
diff is hard to read :-(
Of course feel free to be mega pedantic on all the thing that are not "go friendly", my first go PR :baby_bottle: 

I went with function, don't sure if there is a better way?
Like sometime I need to give the store as [first parameter](https://github.com/LedgerHQ/bitcoin-keychain/pull/24/files#diff-31a59120b1111834534964583197e5db8672d79895943696e69cb9518fdc57c7R436) it feel a bit weird.
Without inheritance can we do better? :thinking:

**Review by onyb — CHANGES_REQUESTED** [review]

File: pkg/keystore/memory.go [hunk 2]
```diff
@@ -51,340 +48,107 @@ func (s *InMemoryKeystore) Delete(id uuid.UUID) error {
...
-// Reset removes derivations and addresses of a keychain corresponding to a UUID from the keystore.
 func (s *InMemoryKeystore) Reset(id uuid.UUID) error {
-	keychain, ok := s.db[id]
+	meta, ok := s.db[id]
 	if !ok {
 		return ErrKeychainNotFound
 	}
 
-	keychain.ResetKeychainMeta()
+	keystoreReset(meta)
```
[the remark this answers is not in the document]
Comment (CedricCabessa) [reply pkg/keystore/memory.go:57]: no, overkill

**Comment by CedricCabessa** [comment]
(those 2 commit can be squashed)

File: pkg/keystore/store.go [hunk 3]
```diff
@@ -322,3 +360,270 @@ func (m *Meta) ResetKeychainMeta() {
...
+func keystoreMarkAddressAsUsed(s Keystore, id uuid.UUID, address string) error {
+	path, err := s.GetDerivationPath(id, address)
+	if err != nil {
+		return err
+	}
+
+	return s.MarkPathAsUsed(id, path)
+}
+
+func (meta *Meta) keystoreGetAddressesPublicKeys(derivations []DerivationPath) ([]string, error) {
```
Comment (onyb) [inline pkg/keystore/store.go:615]: Nit: can you please make sure you use `m` everywhere to follow the Go convention?

File: pkg/keystore/store.go [hunk 4]
```diff
@@ -322,3 +360,270 @@ func (m *Meta) ResetKeychainMeta() {
...
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (m *Meta) keystoreGetAllObservableAddresses(
+	client bitcoin.CoinServiceClient,
+	id uuid.UUID,
```
Comment (onyb) [inline pkg/keystore/store.go:558]: Unused? Also, check the other functions.

**Review by onyb — APPROVED** [review]
Can you please fix the errors reported by `mage -v lint`? Looks good overall! Congrats on your first Go PR.

**Pull request merged**